mailpoet/lib/WooCommerce/Subscription.php

```diff
@@ -229,60 +237,64 @@ public function handleSubscriberOptin(SubscriberEntity $subscriberEntity, bool $
-      $subscriber->confirmedIp = Helpers::getIP();
-      $subscriber->setExpr('confirmed_at', 'NOW()');
+  private function subscribe(SubscriberEntity $subscriber) {
+    $subscriber->setStatus(SubscriberEntity::STATUS_SUBSCRIBED);
+    if (empty($subscriber->getConfirmedIp()) && empty($subscriber->getConfirmedAt())) {
+      $subscriber->setConfirmedIp(Helpers::getIP());
+      $subscriber->setConfirmedAt(new Carbon());
     }
-    $subscriber->save();
-  }
 
-  private function requireSubscriptionConfirmation(Subscriber $subscriber) {
-    // we need to save the subscriber here since handleSubscriberOptin() sets the source but doesn't save the model.
-    // when we migrate this class to use Doctrine we can probably remove this call to save() as the call to persist() below should be enough.
-    $subscriber->save();
-    $subscriberEntity = $this->subscribersRepository->findOneById($subscriber->id);
+    $this->subscribersRepository->persist($subscriber);
+    $this->subscribersRepository->flush();
+  }
 
-    if ($subscriberEntity instanceof SubscriberEntity) {
-      $subscriberEntity->setStatus(SubscriberEntity::STATUS_UNCONFIRMED);
-      $this->subscribersRepository->persist($subscriberEntity);
-      $this->subscribersRepository->flush();
+  private function requireSubscriptionConfirmation(SubscriberEntity $subscriber) {
+    $subscriber->setStatus(SubscriberEntity::STATUS_UNCONFIRMED);
+    $this->subscribersRepository->persist($subscriber);
+    $this->subscribersRepository->flush();
 
-      try {
-        $this->confirmationEmailMailer->sendConfirmationEmailOnce($subscriberEntity);
-      } catch (\Exception $e) {
-        // ignore errors
-      }
+    try {
+      $this->confirmationEmailMailer->sendConfirmationEmailOnce($subscriber);
+    } catch (\Exception $e) {
+      // ignore errors
     }
   }
 
-  private function updateSubscriberStatus(Subscriber $subscriber) {
-    $ss = Subscriber::findOne($subscriber->id);
-    $segmentsCount = $ss->segments()->count();
+  private function updateSubscriberStatus(SubscriberEntity $subscriber) {
+    $segmentsCount = $subscriber->getSubscribedSegments()->count();
+
     if (!$segmentsCount) {
-      $subscriber->status = SubscriberEntity::STATUS_UNSUBSCRIBED;
-      $subscriber->save();
-      $this->unsubscribesTracker->track($subscriber->id, StatisticsUnsubscribeEntity::SOURCE_ORDER_CHECKOUT);
+      $subscriber->setStatus(SubscriberEntity::STATUS_UNSUBSCRIBED);
+      $this->subscribersRepository->persist($subscriber);
+      $this->subscribersRepository->flush();
+      $this->unsubscribesTracker->track((int)$subscriber->getId(), StatisticsUnsubscribeEntity::SOURCE_ORDER_CHECKOUT);
     }
   }
+
+  private function getOptinSegments(): array {
+    $segments = [];
+    $segmentsIds = (array)$this->settings->get(self::OPTIN_SEGMENTS_SETTING_NAME, []);
+
+    foreach ($segmentsIds as $segmentId) {
+      $segments[] = $this->entityManager->getReference(SegmentEntity::class, $segmentId);
```

Comment:
Why did you use `getReference` here? This looks a little bit unsafe to me. If I am not wrong, `getRefernece` doesn't load the entity from the database, and it's possible to save invalid data with this approach. For example, when a segment in settings is deleted.
When you use `$this->segmentsRepository->findBy(['id' => $segmentIds])`, the code will be simpler and also it's possible to remove `EntityManager` from dependecies.

Reply:
I opted to use `getReference` to avoid hitting the database unnecessarily as we don't need the full entity here. But I missed that it was possible for users to delete segments that are added to the `woocommerce.segments` settings and that this would cause problems as you pointed out. I added a new commit using `findBy()` instead of `getReferece()` and simplifying the code.

Reply:
By the way, do you think we should create a ticket to add some protection against the user deleting segments that are part of `woocommerce.segments`?

Reply:
>By the way, do you think we should create a ticket to add some protection against the user deleting segments that are part of woocommerce.segments?

👍🏻 I think that we can create a behavior that at least notifies a user about it.

Reply:
https://mailpoet.atlassian.net/browse/MAILPOET-4597